Assets/LeapMotionModules/UI/Interaction/InteractionBehaviour.cs

```diff
@@ -0,0 +1,692 @@
+    /// fully positionally locked (X, Y, Z) or if a Joint attached to the Rigidbody
+    /// no longer locks its position (e.g. by being destroyed or disabled).
+    /// </summary>
+    public void RefreshPositionLockedState() {
+      if ((_body.constraints & RigidbodyConstraints.FreezePositionX) > 0
+       && (_body.constraints & RigidbodyConstraints.FreezePositionY) > 0
+       && (_body.constraints & RigidbodyConstraints.FreezePositionZ) > 0) {
+        _isPositionLocked = true;
+        return;
+      }
+      else {
+        Joint[] joints = _body.GetComponents<Joint>();
+        foreach (var joint in joints) {
+          FixedJoint fixedJoint = joint as FixedJoint;
+          if (fixedJoint != null) {
+            _isPositionLocked = true; return;
+          }
+          HingeJoint hinge = joint as HingeJoint;
+          if (hinge != null) {
+            _isPositionLocked = true; return;
+          }
+          // SpringJoint -- no check required, spring joints never fully lock position.
+          CharacterJoint charJoint = joint as CharacterJoint;
+          if (charJoint != null) {
+            _isPositionLocked = true; return;
+          }
+          ConfigurableJoint configJoint = joint as ConfigurableJoint;
+          if (configJoint != null
+            && (configJoint.xMotion == ConfigurableJointMotion.Locked
+              || (configJoint.xMotion == ConfigurableJointMotion.Limited
+                && configJoint.linearLimit.limit == 0F))
+            && (configJoint.yMotion == ConfigurableJointMotion.Locked
+              || (configJoint.yMotion == ConfigurableJointMotion.Limited
+                && configJoint.linearLimit.limit == 0F))
+            && (configJoint.zMotion == ConfigurableJointMotion.Locked
+              || (configJoint.zMotion == ConfigurableJointMotion.Limited
+                && configJoint.linearLimit.limit == 0F))) {
+            _isPositionLocked = true; return;
+          }
+        }
+      }
+      _isPositionLocked = false;
+    }
+
+    #endregion
+
+    #region Interaction Layers 
+
+    protected enum CollisionMode {
+      Normal,
+      Grasped
+    }
+    protected CollisionMode _collisionMode;
+
+    protected void FixedUpdateCollisionMode() {
+      CollisionMode desiredCollisionMode = CollisionMode.Normal;
+      if (isGrasped) {
+        desiredCollisionMode = CollisionMode.Grasped;
+      }
+      //else if (_dislocatedBrushCounter < DISLOCATED_BRUSH_COOLDOWN || (_CollisionMode != CollisionMode.Normal && _minHandDistance <= 0.0f)) {
+      //  desiredCollisionMode = CollisionMode.SOFT;
+      //}
+
+      if (_collisionMode != desiredCollisionMode) {
+        _collisionMode = desiredCollisionMode;
+        FixedUpdateLayer();
+      }
+
+      Assert.IsTrue((_collisionMode == CollisionMode.Grasped) == isGrasped);
+    }
+
+    protected void FixedUpdateLayer() {
+      int layer;
+      if (_collisionMode != CollisionMode.Normal) {
```

Comment:
since this is an enum, a switch statement with a default case that throws an exception would be great to enforce correct behavior (or noisy failing!) if a new mode is added